### PR TITLE
coalib: Create `get_commit`

### PR DESCRIPTION
**Pull request by arush0311**
This commit introduces `__commit__` variable containing
the sha hash of the current commit or the commit of `HEAD`
at the time of installation

Closes https://github.com/coala/coala/issues/3367

### Checklist

- [x] I have rebased properly. Please see [our tutorial on rebasing](http://coala.io/git#rebasing).
- [x] I have gone through the [commit guidelines](http://coala.io/commit) and I've followed them.
- [x] I have followed the [guidelines for the commit shortlog](http://coala.io/commit#shortlog).
- [x] I have followed the [guidelines for the commit body](http://coala.io/commit#commit-body).
- [x] I have included the issue URL with the 'Fixes' keyword if the commit fixes a *real bug* and the 'Closes' keyword if it adds a feature or enhancement. For details, check the [issue reference section in our commit guidelines](http://coala.io/commit#issue-reference).
- [x] I have [run all the tests](http://api.coala.io/en/latest/Developers/Executing_Tests.html) and they all pass. If any CI (below) is not green, please fix them. If GitMate issues appear you will have to amend your commits, pushing new commits on top is not sufficient!

### Reviewers
@Makman2

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -62,6 +63,9 @@ def run(self):
               '-b {}'.format(current_version)])
         VERSION = get_version()
 
+with open(COMMIT_FILE, 'w') as cf:
+    cf.write(get_current_commit() + '\n')
+
```
Comment (arush0311) [inline setup.py]: Travis python 2.7 complains is we use `print(get_current_commit, file=cf)` as print is not a function

File: coalib/__init__.py [hunk 2]
```diff
@@ -7,10 +7,46 @@
...
+COMMIT_FILE = join(dirname(__file__), 'COMMIT')
+
+
+def get_current_commit():
+    '''
+    :return:    string containing the current commit sha if coala is
+                initialized as a git repo otherwise an empty string
+    '''
+    try:
+        commit = check_output(['git', 'rev-parse', 'HEAD'])
```
Comment (jayvdb) [inline coalib/__init__.py]: Use `sarge` instead of using `check_output`, as `sarge` provides a sane and safe cross platform interface.

File: coalib/__init__.py [hunk 3]
```diff
@@ -7,10 +7,46 @@
...
 from os.path import join, dirname
+from subprocess import check_output, CalledProcessError
 
 
 VERSION_FILE = join(dirname(__file__), 'VERSION')
 
+COMMIT_FILE = join(dirname(__file__), 'COMMIT')
+
+
+def get_current_commit():
```
Comment (jayvdb) [inline coalib/__init__.py]: this needs to be run in the coalib directory, otherwise it will be wrong.
It is more versatile if the caller specifies the directory that it should be run in, as an argument to this function.

File: coalib/__init__.py [hunk 4]
```diff
@@ -7,10 +7,46 @@
...
+    '''
+    :return:    string containing the current commit sha if coala is
+                initialized as a git repo otherwise an empty string
+    '''
+    try:
```
Comment (jayvdb) [inline coalib/__init__.py]: Before trying to run a subprocess, which is a heavy call, check there is a `.git` directory, which is a quick check.

File: coalib/__init__.py [hunk 5]
```diff
@@ -7,10 +7,46 @@
...
+        commit = ''
+    return commit
+
+
+def get_commit(commit_file):
+    '''
+    :return:    string containing the current commit sha if coala is a git repo
+                otherwise commit sha at the time of installation.
+    '''
+    commit = get_current_commit()
```
Comment (jayvdb) [inline coalib/__init__.py]: specify the coalib directory here.

File: coalib/__init__.py [hunk 6]
```diff
@@ -7,10 +7,46 @@
 
 import sys
 from os.path import join, dirname
+from subprocess import check_output, CalledProcessError
 
 
 VERSION_FILE = join(dirname(__file__), 'VERSION')
```
Comment (jayvdb) [inline coalib/__init__.py]: Please save the `dirname` result, so it can be used multiple times.

File: coalib/__init__.py (hunk 4 above)
Comment (arush0311) [inline coalib/__init__.py]: Actually, the .git is not present coalib, its only present in coala superdirectory, which is not present after the installation, so we cant check for .git folder

File: coalib/__init__.py (hunk 2 above)
Comment (arush0311) [inline coalib/__init__.py]: Why exactly do we need sarge?
IMO sarge donot provide any benifit over check_output in this case
And it also introduces a new dependency. 
Also can we use an external dependency in setup.py?

File: coalib/__init__.py [hunk 7]
```diff
@@ -6,10 +6,61 @@
...
+
+        status = check_output(['git', 'status', '--porcelain'])
+        status = status.decode('utf-8').strip()
+        if status != '':
+            commit += '*'
+    except (OSError, CalledProcessError) as e:
+        commit = ''
+
+    chdir(old_dir)
+    return commit
```
Comment (pratyushprakash) [inline coalib/__init__.py]: Maybe you could use `change_directory` from `coala_utils` here to shorten this up  a bit.

Comment (arush0311) [inline coalib/__init__.py]: actually we may not be able to use external dependencies in setup.py, and ithis function is directly being called by setup.py

**Comment by arush0311** [comment]
@Makman2 can you review?

**Comment by Udayan12167** [comment]
Put the tests in coalaTest

File: coalib/__init__.py (hunk 7 above)
Comment (Makman2) [inline coalib/__init__.py]: > actually we may not be able to use external dependencies in setup.py

why not?

File: coalib/__init__.py [hunk 8]
```diff
@@ -6,10 +6,61 @@
 
 
 import sys
+
```
Comment (Makman2) [inline coalib/__init__.py]: obsolete newline

File: coalib/__init__.py (hunk 7 above)
Comment (arush0311) [inline coalib/__init__.py]: because one of the purpose of setup.py is to install dependencies, and these functions are run before that!

Comment (Makman2) [inline coalib/__init__.py]: ah obviously, nvm then :3

**Comment by arush0311** [comment]
@Udayan12167 Actually, should I do that? the tests do have their own setup and teardown, which donot at all apply to other tests.

**Comment by Udayan12167** [comment]
@arush0311 if you put the methods in the misc directory then you can create a separate file to test them. This way I feel it's difficult to find the tests and could make the test directory cluttered.

**Comment by arush0311** [comment]
@Udayan12167 I understand your point, I will move the tests to coalaTest

**Comment by Udayan12167** [comment]
No I think going the misc directory way is better. Make a `gitHelper.py` in that directory and a parallel `gitHelperTest.py`.

**Comment by arush0311** [comment]
Is it good though? that way we have to keep track of one more file which donot contain external dep and is python2 friendly.

File: setup.py [hunk 9]
```diff
@@ -13,6 +13,8 @@
 
 from coalib import VERSION, assert_supported_version, get_version
 from coalib.misc.BuildManPage import BuildManPage
+from coalib import get_current_commit, COMMIT_FILE, COALIB_DIR
```
Comment (jayvdb) [inline setup.py]: these imports should be near the other imports from the same module.
one way to do that is to move this line up one line.

File: coalib/__init__.py [hunk 10]
```diff
@@ -6,10 +6,60 @@
...
+COALIB_DIR = join(dirname(__file__))
+
+VERSION_FILE = join(COALIB_DIR, 'VERSION')
+
+COMMIT_FILE = join(COALIB_DIR, 'COMMIT')
+
+
+def get_current_commit(git_dir):
+    '''
+    :param git_dir:     string containing the directory to check the
```
Comment (jayvdb) [inline coalib/__init__.py]: 'git_dir' is quite ambiguous.  It could be the directory where `git` is located.

I think this should be more like `git_repo_dir`.

File: coalib/__init__.py [hunk 11]
```diff
@@ -6,10 +6,60 @@
...
+                        current commit in.
+    :return:            string containing the current commit sha if `git_dir`
+                        is initialized as a git repo otherwise an empty string.
+    '''
+    old_dir = getcwd()
+    chdir(git_dir)
```
Comment (jayvdb) [inline coalib/__init__.py]: if we are already in the git repo dir (or any subdirectory), this `chdir` and the one below are not necessary.

File: coalib/__init__.py [hunk 12]
```diff
@@ -6,10 +6,60 @@
...
-VERSION_FILE = join(dirname(__file__), 'VERSION')
+COALIB_DIR = join(dirname(__file__))
+
+VERSION_FILE = join(COALIB_DIR, 'VERSION')
+
+COMMIT_FILE = join(COALIB_DIR, 'COMMIT')
+
+
+def get_current_commit(git_dir):
+    '''
```
Comment (jayvdb) [inline coalib/__init__.py]: coala uses `"""` for docstring start and end.

Comment (jayvdb) [inline coalib/__init__.py]: coala uses `"""` for docstring start and end.

File: coalib/__init__.py [hunk 13]
```diff
@@ -6,10 +6,60 @@
...
+    old_dir = getcwd()
+    chdir(git_dir)
+
+    try:
+        commit = check_output(['git', 'rev-parse', 'HEAD'])
+        commit = commit.decode('utf-8').strip()
+
+        status = check_output(['git', 'status', '--porcelain'])
+        status = status.decode('utf-8').strip()
+        if status != '':
```
Comment (jayvdb) [inline coalib/__init__.py]: `if status`

File: coalib/__init__.py [hunk 14]
```diff
@@ -21,6 +71,10 @@ def get_version():
 __version__ = VERSION
 
 
+COMMIT = get_commit(COALIB_DIR, COMMIT_FILE)
```
Comment (jayvdb) [inline coalib/__init__.py]: this is a large startup overhead for the properly installed package, which is not a git repo.
The overhead for the properly installed package should be as **small** as possible.  No `chdir`; no `subprocess`.
The overhead for the development package, in a git repo, can be large, because it is not the production version.

**Comment by arush0311** [comment]
@Makman2 please review :D

File: coalib/__init__.py [hunk 15]
```diff
@@ -6,10 +6,69 @@
...
+            commit += '*'
+    except (OSError, CalledProcessError) as e:
+        commit = ''
+
+    if(cwd_changed):
+        chdir(old_dir)
+    return commit
+
+
+def get_commit(git_repo_dir, commit_file):
```
Comment (SanketDG) [inline coalib/__init__.py]: why not just use the `COMMIT_FILE` variable?

Comment (arush0311) [inline coalib/__init__.py]: makes it easier to test.

File: setup.py [hunk 16]
```diff
@@ -5,15 +5,21 @@
...
+import setuptools.command.install
+
 from setuptools import find_packages, setup
 from setuptools.command.test import test as TestCommand
 
 from coalib import VERSION, assert_supported_version, get_version
+from coalib import get_current_commit, COMMIT_FILE, COALIB_DIR
 from coalib.misc.BuildManPage import BuildManPage
 
+
```
Comment (Makman2) [inline setup.py]: unrelated change --> separate commit :)

File: coalib/__init__.py [hunk 17]
```diff
@@ -6,10 +6,69 @@
...
+                            of installation.
+    :return:                string containing the current commit sha if
+                            `git_repo_dir` is a git repo otherwise commit sha
+                            in commit_file.
+    """
+    try:
+        with open(commit_file) as cf:
+            commit = cf.readline().strip()
+    # Not using FileNotFoundError, as python2 CI starts complaining
+    except (OSError, IOError):
```
Comment (Makman2) [inline coalib/__init__.py]: ah damn... we need to file an issue to get `FileNotFoundError` there after merge :)

File: coalib/__init__.py [hunk 18]
```diff
@@ -6,10 +6,69 @@
...
+
+def get_commit(git_repo_dir, commit_file):
+    """
+    :param git_repo_dir:    string containing the directory to check the
+                            current commit in.
+    :param commit_file:     filename containing commit sha of HEAD at the time
+                            of installation.
+    :return:                string containing the current commit sha if
+                            `git_repo_dir` is a git repo otherwise commit sha
+                            in commit_file.
```
Comment (Makman2) [inline coalib/__init__.py]: You should start with upper-case letters + Adjust alignment ;)
would recommend you to use this format btw:
```
:param x:
    4-space indented description
:return:
    And again.
```

File: coalib/__init__.py [hunk 19]
```diff
@@ -6,10 +6,69 @@
...
+COMMIT_FILE = join(COALIB_DIR, 'COMMIT')
+
+
+def get_current_commit(git_repo_dir):
+    """
+    :param git_repo_dir:    string containing the directory to check the
+                            current commit in.
+    :return:                string containing the current commit sha if
+                            `git_repo_dir`is initialized as a git repo
+                            otherwise an empty string.
```
Comment (Makman2) [inline coalib/__init__.py]: Start sentences with upper-case + Alignment (see below for recommendation on how to format docs)

File: coalib/__init__.py [hunk 20]
```diff
@@ -6,10 +6,69 @@
...
+
+COMMIT_FILE = join(COALIB_DIR, 'COMMIT')
+
+
+def get_current_commit(git_repo_dir):
+    """
+    :param git_repo_dir:    string containing the directory to check the
+                            current commit in.
+    :return:                string containing the current commit sha if
+                            `git_repo_dir`is initialized as a git repo
```
Comment (Makman2) [inline coalib/__init__.py]: missing space before `is`

File: coalib/__init__.py [hunk 21]
```diff
@@ -6,10 +6,69 @@
...
+                            otherwise an empty string.
+    """
+    old_dir = getcwd()
+    cwd_changed = False
+
+    if(old_dir != git_repo_dir):
```
Comment (Makman2) [inline coalib/__init__.py]: unnecessary braces

File: coalib/__init__.py (hunk 15 above)
Comment (Makman2) [inline coalib/__init__.py]: I think it makes sense to inline the `COMMIT_FILE` and `COALIB_DIR` variables.

Do you know the `unittest.patch` decorator? :)
It allows you to patch fields in a different module, e.g.
```python
@unittest.patch('coalib.COMMIT_FILE', '/home/xyz')
def test_get_commit(self):
    ...
```

File: coalib/__init__.py [hunk 22]
```diff
@@ -6,10 +6,69 @@
...
+                            in commit_file.
+    """
+    try:
+        with open(commit_file) as cf:
+            commit = cf.readline().strip()
+    # Not using FileNotFoundError, as python2 CI starts complaining
+    except (OSError, IOError):
+        commit = ''
+
+    if(commit == 'develop'):
```
Comment (Makman2) [inline coalib/__init__.py]: maybe return the git sha when there's just no `COMMIT` file present?

File: setup.py (hunk 16 above)
Comment (arush0311) [inline setup.py]: What issue should I refer :D ? imo it seems very trivial

File: coalib/__init__.py (hunk 22 above)
Comment (arush0311) [inline coalib/__init__.py]: as @jayvdb said it may increase production overhead. No commit file means the repo was not installed from a git repo, so there will be no git sha..

File: coalib/__init__.py (hunk 15 above)
Comment (arush0311) [inline coalib/__init__.py]: you mean make them one variable? or remove them from parameters?

File: coalib/__init__.py (hunk 18 above)
Comment (arush0311) [inline coalib/__init__.py]: noted :D. as for the case most of the description started with `string`. should I capitalize that?(or perhaps use `str`?)

File: coalib/__init__.py (hunk 15 above)
Comment (arush0311) [inline coalib/__init__.py]: https://github.com/coala/coala/pull/3595#discussion_r96984596

File: coalib/__init__.py (hunk 22 above)
Comment (arush0311) [inline coalib/__init__.py]: https://github.com/coala/coala/pull/3595#discussion_r98343215

File: setup.py (hunk 16 above)
Comment (Makman2) [inline setup.py]: no issue, just a new commit

I know this is trivial, but for the sake of making commits as atomic as possible and ease review (even a bit) we should do it :)

File: coalib/__init__.py (hunk 18 above)
Comment (Makman2) [inline coalib/__init__.py]: If you use `str` you can should enclose it in double backticks then, as you refer to the actual type. So capitalize or wrap `str` in double backticks (for sentences that begin with "string")

File: coalib/__init__.py (hunk 22 above)
Comment (Makman2) [inline coalib/__init__.py]: hmm... imo people install from pip or from repository. If they use e.g. a tar package or similar, the commit file should be included there as well (as it's a packed release normally). I can't imagine what @jayvdb wanted to point out...

Comment (arush0311) [inline coalib/__init__.py]: Sorry the commit file will always be there even if someone donot install from git repo. Exception handling is there just in case

Comment (arush0311) [inline coalib/__init__.py]: should probably change that behaviour though

File: .gitignore [hunk 23]
```diff
@@ -36,3 +36,4 @@ node_modules
 .vendor
 Goopfile.lock
 venv/
+COMMIT
```
Comment (arush0311) [inline .gitignore]: @Makman2 Is this okay? while doing a release one commit file will have to be manually placed

File: coalib/__init__.py (hunk 15 above)
Comment (Makman2) [inline coalib/__init__.py]: ah okay thx for referring to that comment, keep it then :)

File: coalib/__init__.py (hunk 22 above)
Comment (Makman2) [inline coalib/__init__.py]: I think easiest is to have no `COMMIT` file unless an installation (not develop-installation!). If there's no `COMMIT` file, try to get the git hash, if that does not work, return `''` :+1:

@arush0311 what do you think? :)

File: .gitignore (hunk 23 above)
Comment (Makman2) [inline .gitignore]: I thought you generate it during install?

Comment (arush0311) [inline .gitignore]: Yes but that wont happen during installation of released version, as there is no git directory

File: coalib/__init__.py (hunk 22 above)
Comment (arush0311) [inline coalib/__init__.py]: @makman2 during installation of released version there will be no commit file, unless its manually placed before release. In that case the overhead of release version will be more.

File: setup.py [hunk 24]
```diff
@@ -23,11 +27,28 @@
...
+    """
+    Creates COMMIT file with `HEAD` sha if `COALIB_DIR` is a git repo.
+    """
+    current_commit = get_current_commit(COALIB_DIR)
+    if(current_commit != ''):
+        with open(os.path.join(COALIB_DIR, COMMIT_FILE), 'w') as cf:
+            cf.write(get_current_commit(COALIB_DIR) + '\n')
+
+
+class SdistCommand(setuptools.command.sdist.sdist):
```
Comment (arush0311) [inline setup.py]: sure gitmate :stuck_out_tongue:

**Comment by arush0311** [comment]
@Mixih please review if you have time..

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/3595/commits/e0295d2c668c7b60a0c262ce6bbd5a894f851a9a

**Comment by sils** [comment]
are we sure we wanna do IO on every import of our init py? Performance IS a problem with coala and IMO this isn't important enough to make it worse. On installation? Good. On every run? Not so sure.

**Comment by arush0311** [comment]
what about we remove the commit variable and only use get_commit function ?



  <https://mailtrack.io/> Sent with Mailtrack
<https://mailtrack.io/install?source=signature&lang=en&referral=arushgyl@gmail.com&idSignature=22>

**Comment by arush0311** [comment]
@sils @Mixih ^

**Comment by sils** [comment]
sounds good

**Comment by sils** [comment]
sounds good

**Comment by nemaniarjun** [comment]
ack b2324f0fdd076d76b5362237b967994e01406053

**Comment by nemaniarjun** [comment]
@jayvdb there are some changes requested by you, can you look into them once? :D

**Comment by Mixih** [comment]
ping @jayvdb your opinion please.

File: coalib/__init__.py [hunk 25]
```diff
@@ -6,10 +6,69 @@
...
+    :return:
+        String containing the current commit sha if `git_repo_dir` is a
+        git repo otherwise commit sha in commit_file.
+    """
+    try:
+        with open(commit_file) as cf:
+            commit = cf.readline().strip()
+    # Not using FileNotFoundError, as python2 CI starts complaining
+    except (OSError, IOError):
+        commit = ''
```
Comment (jayvdb) [inline coalib/__init__.py]: `commit = None`.  Do not use the empty string unless it is actually an empty string that you need.

File: coalib/__init__.py [hunk 26]
```diff
@@ -6,10 +6,69 @@
...
+
+    try:
+        commit = check_output(['git', 'rev-parse', 'HEAD'])
+        commit = commit.decode('utf-8').strip()
+
+        status = check_output(['git', 'status', '--porcelain'])
+        status = status.decode('utf-8').strip()
+        if status:
+            commit += '*'
+    except (OSError, CalledProcessError) as e:
```
Comment (jayvdb) [inline coalib/__init__.py]: why catch these?
When these exceptions occur, this method is *not* returning a commit representation, which means it has failed, so it should be throwing an exception.

File: setup.py [hunk 27]
```diff
@@ -52,6 +75,14 @@ def run(self):
         sys.exit(errOne or errTwo)
 
 
+class DevelopCommand(setuptools.command.develop.develop):
+
+    def run(self):
+        with open(os.path.join(COALIB_DIR, COMMIT_FILE), 'w') as cf:
+            cf.write('develop\n')
```
Comment (jayvdb) [inline setup.py]: what is the benefit of writing 'develop' into the file.  Isn't that the same as no commit file?
The COMMIT file is mandatory because it is listed in `package_data`?

Maybe after running `setuptools.command.develop.develop.run(..)`, remove the `COMMIT` file again...?

File: setup.py [hunk 28]
```diff
@@ -23,11 +28,29 @@
 assert_supported_version()
 
 
+def create_commit_file():
+    """
+    Creates COMMIT file with `HEAD` sha if `COALIB_DIR` is a git repo.
+    """
+    current_commit = get_current_commit(COALIB_DIR)
+    if(current_commit != ''):
```
Comment (jayvdb) [inline setup.py]:
```diff
-if(current_commit != '')
+if (current_commit != '')
```

but ... wtf...

If you can not get a commit identifier, it is probably because the user is installing from a tarball.  You must write a COMMIT file so that the package doesnt try to run `git` at startup every time the package is loaded.

File: coalib/__init__.py (hunk 17 above)
Comment (arush0311) [inline coalib/__init__.py]: yeah its not needed now, previously it was needed when we were using `__commit__` variable

**Comment by SanketDG** [comment]
unack b2324f0

**Comment by Mixih** [comment]
I think we've drifted a bit from the original goal. Can we have a discussion on the issue please?

File: setup.py [hunk 29]
```diff
@@ -72,14 +76,28 @@ def run(self):
...
 
+def _get_public_version_string(scm_version):
+    return VERSION
+
+
+def _get_local_version_string(scm_version):
+    if 'dev' in VERSION:
+        return get_local_node_and_date(scm_version)
+    else:
+        return ''
```
Comment (shreyans800755) [inline setup.py]: Can be done in a single line like `return get_local_node_and_date(scm_version) if 'dev' in VERSION else ''`

File: .misc/check_setuptools.py [hunk 30]
```diff
@@ -2,14 +2,16 @@
...
 
 import setuptools
 
 
 def get_setuptools_version():
     with open('requirements.txt') as f:
         for line in f:
-            if line.startswith('setuptools'):
+            # Ensure that line matched is setuptools
+            if re.match(r'^setuptools(>|<|~|=|$)', line):
```
Comment (shreyans800755) [inline .misc/check_setuptools.py]: `re.compile` is better option since we need to match it multiple times and regex remains same ever time.

**Comment by Makman2** [comment]
@arush0311 why closing? 😞

**Comment by arush0311** [comment]
Hey! I am sorry.. I don't think this pr is relevant anymore and i thought it might be wasting time of reviewers and I keep getting spammed by gitmate

**Comment by Makman2** [comment]
I think it's still good to have :) However if you don't want to continue, it's fine :)
